Source/Particles/ParticleCreation/SmartCopy.H

```diff
@@ -0,0 +1,219 @@
+#ifndef SMART_COPY_H_
+#define SMART_COPY_H_
+
+#include <DefaultInitialization.H>
+
+#include <AMReX_GpuContainers.H>
+#include <AMReX_ParallelDescriptor.H>
+
+#include <map>
+#include <string>
+
+using NameMap = std::map<std::string, int>;
+using PolicyVec = amrex::Gpu::DeviceVector<InitializationPolicy>;
+
+struct SmartCopyTag
+{
+    std::vector<std::string> common_names;
+    amrex::Gpu::DeviceVector<int> src_comps;
+    amrex::Gpu::DeviceVector<int> dst_comps;
+
+    int size () const noexcept { return common_names.size(); }
+};
+
+PolicyVec getPolicies (const NameMap& names) noexcept;
+
+SmartCopyTag getSmartCopyTag (const NameMap& src, const NameMap& dst) noexcept;
+
+/**
+ * \brief Sets the ids of newly created particles to the next values.
+ *
+ * \tparam PTile the particle tile type
+ *
+ * \param ptile the particle tile
+ * \param old_size the index of the first new particle
+ * \param num_added the number of particles to set the ids for.
+ */
+template <typename PTile>
+void setNewParticleIDs (PTile& ptile, int old_size, int num_added)
+{
+    int pid;
+#ifdef _OPENMP
+#pragma omp critical (ionization_nextid)
+#endif
+    {
+        pid = PTile::ParticleType::NextID();
+        PTile::ParticleType::NextID(pid + num_added);
+    }
+
+    const int cpuid = amrex::ParallelDescriptor::MyProc();
+    auto pp = ptile.GetArrayOfStructs()().data() + old_size;
+    amrex::ParallelFor(num_added, [=] AMREX_GPU_DEVICE (int ip) noexcept
+    {
+        auto& p = pp[ip];
+        p.id() = pid+ip;
+        p.cpu() = cpuid;
+    });
+}
+
+/**
+ * \brief This is a functor for performing a "smart copy" that works
+ * in both host and device code.
+ *
+ * A "smart" copy does the following. First, the destination particle
+ * components are initialized to the default values for that component
+ * type. Second, if a given component name is found in both the src
+ * and the dst, then the src value is copied.
+ *
+ * Particle structs - positions and id numbers  - are always copied.
+ *
+ * You don't create this directly - use the SmartCopyFactory object below.
+ */
+struct SmartCopy
+{
+    int m_num_copy_real;
+    const int* m_src_comps_r;
+    const int* m_dst_comps_r;
+
+    int m_num_copy_int;
+    const int* m_src_comps_i;
+    const int* m_dst_comps_i;
+
+    const InitializationPolicy* m_policy_real;
+    const InitializationPolicy* m_policy_int;
+
+    template <typename DstData, typename SrcData>
+    AMREX_GPU_HOST_DEVICE AMREX_FORCE_INLINE
+    void operator() (DstData& dst, const SrcData& src, int i_src, int i_dst) const noexcept
+    {
+        // the particle struct is always copied over
+        dst.m_aos[i_dst] = src.m_aos[i_src];
+
+        // initialize the real components
```

Comment:
general questions: do we have to initialize all dst real/int components are can we potentially just init the ones we did not copy?

Reply:
We don't have to, but I think the logic would be kinda complicated and probably not appreciably faster.

Reply:
I agree with @atmyers

Reply:
Makes sense, no need for premature optimizations.